File: swex/macro-feature/link-external-file/csharp/LinkFeatureToExternalFile/LinkFileMacroFeature.cs
using CodeStack.SwEx.MacroFeature;
using CodeStack.SwEx.MacroFeature.Attributes;
using System;
using System.Linq;
using System.Runtime.InteropServices;
using CodeStack.SwEx.MacroFeature.Base;
using SolidWorks.Interop.sldworks;
using CodeStack.SwEx.Common.Attributes;
using CodeStack.SwEx.Examples.LinkFeatureToExternalFile.Properties;

namespace CodeStack.SwEx.Examples.LinkFeatureToExternalFile
{

    [ComVisible(true), Guid("3590F6C4-6345-4A0A-9161-196896F3B335")]
    [ProgId("CodeStack.LinkFileMacroFeature.CSharp")]
    [Options("LinkFileMacroFeature")]
    [Icon(typeof(Resources), nameof(Resources.linked_part_icon))]
    public class LinkFileMacroFeature : MacroFeatureEx<LinkFileMacroFeatureParameters, LinkFileMacroFeatureHandler>
    {
        protected override MacroFeatureRebuildResult OnRebuild(LinkFileMacroFeatureHandler handler, LinkFileMacroFeatureParameters parameters)
        {
            if (handler.LastError == null)
            {
                parameters.FileLastUpdateTimeStamp = handler.LastUpdateStamp;

                SetParameters(handler.Model, handler.Feature, handler.Feature.GetDefinition() as IMacroFeatureData, parameters);

                return MacroFeatureRebuildResult.FromBodies(
                    handler.CachedBodies.Select(b => b.ICopy()).ToArray(), 
                    handler.Feature.GetDefinition() as IMacroFeatureData);
            }
            else
            {
                return MacroFeatureRebuildResult.FromStatus(false, handler.LastError.ToString());
            }
        }
    }
}

File: swex/macro-feature/link-external-file/vb-net/LinkFeatureToExternalFile/LinkFileMacroFeature.vb
Imports CodeStack.SwEx.MacroFeature
Imports CodeStack.SwEx.MacroFeature.Attributes
Imports System.Runtime.InteropServices
Imports CodeStack.SwEx.MacroFeature.Base
Imports SolidWorks.Interop.sldworks
Imports CodeStack.SwEx.Common.Attributes
Imports CodeStack.SwEx.Examples.LinkFeatureToExternalFile.My.Resources

<ComVisible(True), Guid("11F925DE-805F-40F6-AF26-16C760234E82")>
<ProgId("CodeStack.LinkFileMacroFeature.VBNET")>
<Options("LinkFileMacroFeature")>
<Icon(GetType(Resources), NameOf(Resources.linked_part_icon))>
Public Class LinkFileMacroFeature
	Inherits MacroFeatureEx(Of LinkFileMacroFeatureParameters, LinkFileMacroFeatureHandler)

	Protected Overrides Function OnRebuild(ByVal handler As LinkFileMacroFeatureHandler, ByVal parameters As LinkFileMacroFeatureParameters) As MacroFeatureRebuildResult
		If handler.LastError Is Nothing Then
			parameters.FileLastUpdateTimeStamp = handler.LastUpdateStamp
			SetParameters(handler.Model, handler.Feature, TryCast(handler.Feature.GetDefinition(), IMacroFeatureData), parameters)
			Return MacroFeatureRebuildResult.FromBodies(handler.CachedBodies.Select(Function(b) b.ICopy()).ToArray(), TryCast(handler.Feature.GetDefinition(), IMacroFeatureData))
		Else
			Return MacroFeatureRebuildResult.FromStatus(False, handler.LastError.ToString())
		End If
	End Function

End Class
